src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -826,9 +857,32 @@ private[internal] trait TypeMaps {
     }
   }
 
+  /** A map to implement the `substSym` method. */
+  class SubstSymMap(override val from: List[Symbol], override val to: List[Symbol]) extends AbstractSubstSymMap
+
+  class MutableSubstSymMap(var _from: List[Symbol], private var _to: List[Symbol]) extends AbstractSubstSymMap {
```

Comment:
I really, really dislike this.

Reply:
Well, It did not work. Turns out that the `cloneSymbols` method is indeed reentrant... that is, there is a loop from `cloneSymbols` to `cloneSymbols`, through some classes. 

```
	at scala.reflect.internal.Symbols.cloneSymbols(Symbols.scala:3744)
	at scala.reflect.internal.Symbols.cloneSymbols$(Symbols.scala:3740)
	at scala.reflect.internal.SymbolTable.cloneSymbols(SymbolTable.scala:28)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:140)
	at scala.reflect.internal.Types$MethodType.mapOver(Types.scala:2922)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstMap.apply(TypeMaps.scala:744)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstSymMap.apply(TypeMaps.scala:805)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.applyToSymbolInfo(TypeMaps.scala:123)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.loop$1(TypeMaps.scala:117)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.firstChangedSymbol(TypeMaps.scala:121)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:135)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:128)
	at scala.reflect.internal.Types$RefinedType.mapOver(Types.scala:1864)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstMap.apply(TypeMaps.scala:744)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstSymMap.apply(TypeMaps.scala:805)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstSymMap.apply(TypeMaps.scala:790)
	at scala.collection.immutable.List.loop$3(List.scala:445)
	at scala.collection.immutable.List.mapConserve(List.scala:470)
	at scala.reflect.internal.Types$RefinedType.mapOver(Types.scala:1863)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstMap.apply(TypeMaps.scala:744)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstSymMap.apply(TypeMaps.scala:805)
	at scala.reflect.internal.Types$TypeBounds.mapOver(Types.scala:1579)
	fat scala.reflect.internal.tpe.TypeMaps$AbstractSubstMap.apply(TypeMaps.scala:744)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstSymMap.apply(TypeMaps.scala:805)
	at scala.reflect.internal.tpe.TypeMaps$AbstractSubstSymMap.apply(TypeMaps.scala:790)
	at scala.reflect.internal.Symbols$Symbol.modifyInfo(Symbols.scala:1574)
	at scala.reflect.internal.Symbols.$anonfun$cloneSymbols$3(Symbols.scala:3748)
	at scala.collection.immutable.List.foreach(List.scala:306)
	at scala.reflect.internal.Symbols.cloneSymbols(Symbols.scala:3748)
	at scala.reflect.internal.Symbols.cloneSymbols$(Symbols.scala:3740)
	at scala.reflect.internal.SymbolTable.cloneSymbols(SymbolTable.scala:28)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:140)
	at scala.reflect.internal.Types$ExistentialType.mapOver(Types.scala:3211)
```

So I had to drop the mutable single instance with two mutable variables of lists. Instead, now I am using two mutable lists of immutable lists (still testing it...)